tests/e2e/specs/saved-cards/saved-cards-checkout.spec.js

```diff
@@ -0,0 +1,119 @@
+/**
+ * External dependencies
+ */
+import config from 'config';
+
+/**
+ * Internal dependencies
+ */
+import { CustomerFlow, uiUnblocked } from '../../utils';
+import {
+	fillCardDetails,
+	confirmCardAuthentication,
+} from '../../utils/payments';
+
+const TIMEOUT = 100000;
+
+const cards = [
+	[ 'basic', config.get( 'cards.basic' ) ],
+	[ '3DS', config.get( 'cards.3ds' ) ],
+];
+
+describe( 'Saved cards ', () => {
+	describe.each( cards )(
+		'when using a %s card added on checkout',
+		( cardType, card ) => {
+			beforeAll( async () => {
+				// Increase default value to avoid test failing due to timeouts.
+				page.setDefaultTimeout( 30000 );
+				// running the login flow takes more than the default timeout of 5 seconds,
+				// so we need to increase it to run the login in the beforeAll hook
+				jest.setTimeout( TIMEOUT );
+				await CustomerFlow.login();
+			} );
+
+			afterAll( async () => {
+				await CustomerFlow.logout();
+			} );
+
+			it(
+				'should save the card',
+				async () => {
+					await setupProductCheckout();
+					await CustomerFlow.selectNewPaymentMethod();
+					await fillCardDetails( page, card );
+					await CustomerFlow.toggleSavePaymentMethod();
+
+					if ( 'basic' === cardType ) {
+						await CustomerFlow.placeOrder();
+					} else {
+						await expect( page ).toClick( '#place_order' );
+						await confirmCardAuthentication( page, cardType );
+						await page.waitForNavigation( {
+							waitUntil: 'networkidle0',
+						} );
+					}
+
+					await expect( page ).toMatch( 'Order received' );
+
+					// validate that the payment method has been added to the customer.
+					await CustomerFlow.goToPaymentMethods();
+					await expect( page ).toMatch( card.label );
+					await expect( page ).toMatch(
+						`${ card.expires.month }/${ card.expires.year }`
+					);
+				},
+				TIMEOUT
+			);
+
+			it(
+				'should process a payment with the saved card',
+				async () => {
+					await setupProductCheckout();
+					await CustomerFlow.selectSavedPaymentMethod(
+						`${ card.label } (expires ${ card.expires.month }/${ card.expires.year })`
+					);
+
+					if ( 'basic' === cardType ) {
+						await CustomerFlow.placeOrder();
+					} else {
+						await expect( page ).toClick( '#place_order' );
+						await confirmCardAuthentication( page, cardType );
+						await page.waitForNavigation( {
+							waitUntil: 'networkidle0',
+						} );
+					}
+
+					await expect( page ).toMatch( 'Order received' );
+				},
+				TIMEOUT
+			);
+
+			it(
+				'should delete the card',
+				async () => {
+					await CustomerFlow.goToPaymentMethods();
+					await CustomerFlow.deleteSavedPaymentMethod( card.label );
+					await expect( page ).toMatch( 'Payment method deleted' );
```

Comment:
We could go and check that the card is not listed on checkout page anymore as well. How do you think is it necessary?

Reply:
We could, but that could cause some issues if/when the tests are run concurrently. Let's say for example that the saved card is added in my account and checkout tests. Even though it was deleted in one of the tests, there could still be an entry for the card that's being used in the other test.

That can be easily fixed by adding a second basic card, but I'm not sure if it's necessary, since the `Payment method deleted` message wouldn't appear if the card was not deleted successfully. What do you think?

Reply:
Fair point, I agree. Check for deleted message should be enough.